infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogEntrySinkBufferManager.java

```diff
@@ -74,4 +75,24 @@ public boolean verifyMessageType(LogReplicationEntry entry) {
 
         return true;
     }
+
+    public void processBuffer() {
```

Comment:
In your case, when it receives 5, it will process 5 and right after process it:
1. it will update the lastSeqNum
2. it will look at the buffer to see if there are any messages are in order to be processed.

Reply:
the map is indexed according to the preTs/preSeq number. 
For your case the map: 
after process message 2, the lastTs = 2
for message 7, the preTs is 5
for message 8, the preTs is 7
for message 5, the preTs is 2: we will process it and update the lastTs as 5, use 5 as the index to lookup and find message 7.